### PR TITLE
msi: migrate td-agent*.bat

### PR DESCRIPTION
**Pull request by kenhys**
msi: migrate td-agent*.bat

Before:

c:\opt\td-agent\bin\td-agent.bat
c:\opt\td-agent\bin\td-agent-gem.bat

After:

c:\opt\td-agent\fluentd.bat
c:\opt\td-agent\fluent-gem.bat

If td-agent-gem.bat was renamed to fluent-gem.bat, it conflict with
c:\opt\td-agent\fluent-gem.bat which fluentd gem provides.
Not to cause such a situation, put them under c:\opt\td-agent.

To keep compatibility, td-agent.bat and td-agent-gem are provided
as hardlink.

### CODE REVIEW

**Comment by kenhys** [comment]
failure:: assets/fluentd.bat is overriden.

**Review by daipom — APPROVED** [review]
These diffs look good to me.
About the link files, I will confirm them after the installer is built.

**Comment by ashie** [comment]
> So in this PR, changed to provide unique batch file - fluent.bat.

`fluent` means to wide nowaday. Although it's origianl name of fluentd, it means our organization name.
So that it might be ambiguous for users what the `fluent.bat` does.

How about placing `fluentd.bat` and `fluent-gem.bat` just under `/opt/td-agent(fluent)`, and add there to `$PATH`.
I think it has some merits.

* Placing execution files just under the application folder is quite usual on Windows.
* Although they have same name with the one under `/opt/td-agent/bin`, users can  easily recognize that the files just under `/opt/td-agent/` have higher priority.
* They won't overwritten by gem.

Or leaving td-agent.bat and td-agent-gem.bat as is and put symlinks with `fluent` prefix just under `/opt/td-aget(fluent)` might be better.

**Comment by daipom** [comment]
@ashie  I see. I agree.

@kenhys About the link files:

```console
$ td-agent.lnk --version
td-agent 5.0.0 fluentd 1.16.1 (0a6d706a9cee5882d751b2cc6169696709df0134)

$ td-agent --version
'td-agent' は、内部コマンドまたは外部コマンド、
操作可能なプログラムまたはバッチ ファイルとして認識されていません。
```

**Comment by daipom** [comment]
It seems that we can omit `.bat` but we can't omit `.lnk`.
I don't know this Windows specification well though,,,

**Comment by ashie** [comment]
> NOTE: There is no need to provide td-agent-gem.bat as asset
> because it is almost same as fluent-gem.bat which is installed by
> fluentd gem.

It might be better to modify to set `GEM_HOME` and `GEM_PATH` like the one on GNU/Linux.
Setting them can suppress referring other directories unexpectedly configured by users.

**Comment by kenhys** [comment]
> $ td-agent --version
'td-agent' は、内部コマンドまたは外部コマンド、
操作可能なプログラムまたはバッチ ファイルとして認識されていません。

Thank you for verifying it.

It seems that hardlink is better solution...

**Comment by kenhys** [comment]
> How about placing fluentd.bat and fluent-gem.bat just under /opt/td-agent(fluent), and add there to $PATH.
I think it has some merits.

Putting these files under /opt/td-agent is a good approach. I'll follow it.

**Comment by kenhys** [comment]
Even though without custom action, rollback occurs. checking...

**Comment by kenhys** [comment]
> Even though without custom action, rollback occurs. checking...

[withoutca.log](https://github.com/fluent/fluent-package-builder/files/11561514/withoutca.log)

Terminated after property change:

> MSI (c) (88:1C) [14:50:00:813]: PROPERTY CHANGE: Adding dir8D9828541E9745DA0C06D1FC23D6C150 property. Its value is 'c:\opt\

File: fluent-package/msi/source.wxs [hunk 1]
```diff
@@ -170,18 +172,71 @@
...
                   BinaryKey="WixCA"
                   DllEntry="WixQuietExec64"
                   Execute="deferred"
                   Return="check"
                   Impersonate="no" />
+
+    <Property Id="CreateCompatTdAgentBat" Value=" "/>
+    <CustomAction Id="SetCreateCompatTdAgentBat"
+                  Property="CreateCompatTdAgentBat"
+                  Value="&quot;cmd&quot; /c &quot;fsutil hardlink create[PROJECTLOCATION]bin\td-agent.bat [PROJECTLOCATION]fluentd.bat&quot;"/>
```
Comment (daipom) [inline fluent-package/msi/source.wxs:186]:
```suggestion
                  Value="&quot;cmd&quot; /c &quot;fsutil hardlink create [PROJECTLOCATION]bin\td-agent.bat [PROJECTLOCATION]fluentd.bat&quot;"/>
```

**Comment by kenhys** [comment]
Old bat is provided as hardlink.

image

**Comment by kenhys** [comment]
checking with/without ComponentGuidGenerationSeed.

**Review by ashie — CHANGES_REQUESTED** [review]
Could you modify also td-agent-prompt.bat to add top directory to `PATH`? The top folder should have higher priority than `bin`.

https://github.com/fluent/fluent-package-builder/blob/0d64c2819f22a129ab9023871b34cb9b911048d7/fluent-package/msi/assets/fluent-package-prompt.bat#L2

**Comment by kenhys** [comment]
dead hardlink issue was fixed.

**Comment by kenhys** [comment]
TODO: fix mismatched path issue

**Comment by kenhys** [comment]
> `TODO:` fix mismatched path issue

Fixed it.

File: fluent-package/msi/assets/fluent-package-prompt.bat [hunk 2]
```diff
@@ -1,3 +1,4 @@
 @echo off
 set PATH="%~dp0\bin";%PATH%
+set PATH="%~dp0";%PATH%
```
Comment (ashie) [inline fluent-package/msi/assets/fluent-package-prompt.bat:3]: `%~dp0\bin` is still needed too.

Comment (ashie) [reply fluent-package/msi/assets/fluent-package-prompt.bat:3]: For `fluent-cat`, `fluent-ctl` and so on.

Comment (kenhys) [reply fluent-package/msi/assets/fluent-package-prompt.bat:3]: > %~dp0\bin is still needed too.

`%~dp0\bin` is not removed in this PR. Am i missing?

Comment (ashie) [reply fluent-package/msi/assets/fluent-package-prompt.bat:3]: Sorry, I overlooked it.

**Review by daipom — APPROVED** [review]
I have confirmed this msi package works on my environment.

On fluent-package-prompt:

* Both `$ fluentd --version` and `$ td-agent --version` work expectedly.
* Both `$ fluent-gem environment` and `$ td-agent-gem environment` work expectedly.
* `$ fluent-ctl restart` works correctly.

On PowerShell:

* `$ Restart-Service fluentdwinsvc` works correctly.

Etc:

* It can be uninstalled correctly.

**Comment by ashie** [comment]
I'm still checking...

**Review by ashie — APPROVED** [review]

**Comment by ashie** [comment]
Thanks!

**Pull request merged**